dpsim/src/PFSolverPowerPolar.cpp

```diff
@@ -353,47 +368,96 @@ Real PFSolverPowerPolar::Q(UInt k) {
 }
 
 void PFSolverPowerPolar::calculatePAndQAtSlackBus() {
-    for (auto k: mVDBusIndices) {
+    for (auto topoNode: mVDBuses) {
+        auto node_idx = topoNode->matrixNodeIndex();
+
+        // calculate power flowing out of the node into the admittance matrix (not into PQ Loads)
         CPS::Complex I(0.0, 0.0);
-        for (UInt j = 0; j < mSystem.mNodes.size(); ++j) {
-            I += mY.coeff(k, j) * sol_Vcx(j);
-        }
-        CPS::Complex S(0.0, 0.0);
-        S = sol_Vcx(k) * conj(I);
-        sol_P(k) = S.real();
-        sol_Q(k) = S.imag();
-        for(auto extnet : mExternalGrids){
-            if(extnet->mPowerflowBusType==CPS::PowerflowBusType::VD){
-			    extnet->updatePowerInjection(S*mBaseApparentPower);
+        for (UInt j = 0; j < mSystem.mNodes.size(); ++j)
+            I += mY.coeff(node_idx, j) * sol_Vcx(j);
+        CPS::Complex S = sol_Vcx(node_idx) * conj(I);
+
+        // GenPower = Power flowing out of node + PQLoadsPower
+        for(auto comp : mSystem.mComponentsAtNode[topoNode])
+            if (auto loadPtr = std::dynamic_pointer_cast<CPS::SP::Ph1::Load>(comp))
+                S += Complex(**(loadPtr->mActivePowerPerUnit), **(loadPtr->mReactivePowerPerUnit));
+
+        // Set power of external network injections of SG (VD)
```

Comment:
@martinmoraga 
`// Set power of either VD-type external network injection or VD-type synchronous generator depending on what is connected`
Is this is what you mean here?